rfc/0017-policy-server-reuse-wasm-optimized-modules.md

```diff
@@ -0,0 +1,294 @@
+[`Lease`](https://kubernetes.io/docs/concepts/architecture/leases/) to ensure
+that only one of the Pods (the leader) has write access to the PV.
+
+## Using Kubernetes `Lease`
+
+The `Lease` resource is a primitive leveraged by Kubernetes controllers to build
+a lock system that is then used for the leader election process.
+
+The hard job of building a lock system is done on the client side. For example,
+the Kubernetes Go client does that inside of the
+[`k8s.io/client-go/tools/leaderelection/resourcelock`](https://pkg.go.dev/k8s.io/client-go@v0.27.3/tools/leaderelection/resourcelock)
+package.
+In the rust ecosystem this functionality is provided by the [kubert](https://crates.io/crates/kubert)
+crate. This crate is also used by one of [linkerd controllers](https://github.com/linkerd/linkerd2/tree/main/policy-controller).
+
+We can rely on [`kubert::lease::LeaseManager`](https://docs.rs/kubert/0.16.1/kubert/lease/struct.LeaseManager.html)
+to ensure exclusive write access to this shared Persistent Volume.
+
+## New deployment model
+
+Policy Server would continue to be deployed using a Deployment object.
+A new Persistent Volume would then be defined. This is where all the `.wasm`
+files and all their optimized versions would be saved.
+
+The deployment would then define a new init container. The Policy Server
+configuration would be injected into the init container as a read-only file.
+The container would also have write access to the Persistent Volume described
+above.
+
+The init container would then run a program that performs the following steps:
+
+1. Obtain ownership of a lock via backed by a Kubernetes Lease resource.
+2. Wait until exclusive ownership is obtained.
+3. Inspect the configuration of Policy Server, download all the missing `.wasm`
+  modules
+4. For each downloaded `.wasm` module, look for an optimized version of it inside
+  of the Persistent Volume
+5. If no optimized version is found, or if the found one is not compatible with
+  the current wasmtime engine, generate a new one and write it to disk
+6. Once all the modules have been downloaded and optimized exit
+
+The typical Policy Server deployment is made by two Pods. In this case, one Pod
+will obtain the exclusive access to the shared lock. Its init container will
+eventually update the contents of the Persistent Volume and then it will start.
+The other Pod will be initially "stuck" with its init container waiting to acquire
+the shared lock. Once the lock is owned, the init container program will exit
+almost immediately, given all the modules and their optimized versions are
+already inside of the Persistent Volume.
+
+The usage of an init container has the following benefits:
+
+* Create a clear representation of the bootstrap process of the Policy Server
+  Pod
+* Avoid further complexity to be added to the Policy Server codebase
+
+The "download & optimize" code would be provided by a dedicated binary that is
```

Comment:
I suppose this phrase is a bit redundant.

Reply:
Fixed, thanks